pkg/controller/nodecsrapprover/node_csr_approver.go

```diff
@@ -59,19 +61,72 @@ type reconciler struct {
 }
 
 func Add(mgr manager.Manager) error {
-	certClient, err := certificatesv1client.NewForConfig(mgr.GetConfig())
+	// TODO: delete whole file node_csr_approver_v1beta1.go and dynamic API groups discovery
+	// after we drop kubernetes 1.18 support
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
 	if err != nil {
-		return fmt.Errorf("failed to create certificate client: %v", err)
+		return fmt.Errorf("failed to init discovery client: %w", err)
 	}
 
-	r := &reconciler{Client: mgr.GetClient(), certClient: certClient.CertificateSigningRequests()}
-	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r})
+	srvGroups, err := discoveryClient.ServerGroups()
+	if err != nil {
+		return fmt.Errorf("failed to server API groups: %w", err)
+	}
+
+	certificatesVersionFound := ""
+	for _, group := range srvGroups.Groups {
+		if group.Name != "certificates.k8s.io" {
+			continue
+		}
+
+		for _, groupVersion := range group.Versions {
+			if groupVersion.Version == "v1" {
+				certificatesVersionFound = "v1"
+			}
+
+			if certificatesVersionFound == "" {
+				certificatesVersionFound = "v1beta1"
+			}
+		}
+	}
+
+	var (
+		rec       reconcile.Reconciler
+		watchType client.Object
+	)
+
+	switch certificatesVersionFound {
+	case "v1":
+		certClient, err := certificatesv1client.NewForConfig(mgr.GetConfig())
+		if err != nil {
+			return fmt.Errorf("failed to create certificate client: %v", err)
+		}
+		rec = &reconciler{Client: mgr.GetClient(), certClient: certClient.CertificateSigningRequests()}
+		watchType = &certificatesv1.CertificateSigningRequest{}
+	case "v1beta1":
```

Comment:
Is it expected that we don't assign `watchType` for v1beta1?

Reply:
good catch, no, it was forgotten